app/demo-geant-integration/RunAction.cc

```diff
@@ -0,0 +1,54 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file demo-geant-integration/RunAction.cc
+//---------------------------------------------------------------------------//
+#include "RunAction.hh"
+
+namespace demo_geant
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with Celeritas setup options and shared data.
+ */
+RunAction::RunAction(SPConstOptions options,
+                     SPParams       params,
+                     SPTransporter  transport)
+    : options_(options), params_(params), transport_(transport)
+{
+    CELER_EXPECT(options_);
+    CELER_EXPECT(params_);
+    CELER_EXPECT(transport_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Initialize Celeritas.
+ */
+void RunAction::BeginOfRunAction(const G4Run* run)
+{
+    CELER_EXPECT(run);
+
+    // Initialize shared data
+    params_->Initialize(*options_);
```

Comment:
Should we initialized this only on the master thread?  - i.e., should param_ be static as it is invariant and read only?

Reply:
Since none of the Celeritas code is static I'd rather not make new static variables... plus it's not clear to me how we can share the params among the various `UserActions` but initialize them only on one. We can only initialize our params inside `BeginOfRunAction` (I think?) since that's the first place after the Geant4 physics has been loaded. The `Initialize` function is set up so that only the first thread that hits it will initialize it. Are you just suggesting to change it to a `static` instead of a `shared_ptr`? If so, would you do that just for the demo app here, or globally? Would that be compatible with other frameworks? I feel like `shared_ptr` gives the most flexibility.

Reply:
Technically, the static variable is in the demo application, not in the celeritas code.   Typically we can do something like

1) add RunAction in `ActionInitialization::BuildForMaster()`
2) initialize the static param_ only once in the master thread inside `RunAction::BeginOfRunAction(...) `

```
    // Initialize shared data                                                        
    if (IsMaster() && run != nullptr)                                          
    {
       params_->Initialize(*options_);
       CELER_ASSERT(*params_);
    }

```
3) probably make transporter shareable static too, so that we can move both param_ and transporter out of `ActionInitialization::Build()` to a new helper class, which may be sharable static too (i.e, global).  Yes, too many statics!  In this way, however, the number of transporters (or vector of transporter) on-the-fly for multiple GPUs would be equivalent to the number of GPUs, rather than the number of CPU threads.  If the transporter is to be thread local, then we may still need mutex anyway when it is invoked.  Of course, it is just one of many approaches and may not the most flexible way.  Anyway, whatever works may be fine for now as this plain vanilla MT approach will not be the final choice.

Reply:
`Initialize` is safe to call multiple times from multiple threads. Can you also guarantee that the master thread will have completed `BeginOfRunAction` before any of the worker threads? And will it work in Serial mode too?

My only problem with using a static variable for the demo app is that it might make it harder to follow when and who should initialize the parameters, and harder to see where the data has to be passed around. (You know, the usual problems with global variables.)

The `LocalTransporter` could also be thread-local, at least for the MT/tasking model. I figured that since CMSSW can't use `thread_local` (because tasks may change threads, etc.) it would be better not to use it for this example.

If you think that static variables and thread-local static data would be easier to understand for Geant4 users than shared pointers, though, I'm happy to change to that method 😅 or perhaps could let you do it in a follow-up commit? I just know that I had a really hard time following the movement of data in the acc01 example.

Reply:
No problem.  Let's make it work as the way that you are comfortable for now.  Regarding to the initialization, I confirm that `BeginOfRunAction` of the master should be completed before worker threads are spawned.  Anyway, can you reproduce the crash at the end of run without setting CELER_DISABLE_DEVICE=1 or is it a problem of my side?

Reply:
It's working with CUDA under the CI... what error(s) are you seeing?

Reply:
@whokion it is also crashing sometimes for me (segfaulting or errors like `malloc_consolidate(): invalid chunk size`, `double free or corruption (fasttop)`...). Is that similar to what you are seeing?

Reply:
My case, from
#17 celeritas::activate_device(celeritas::Device&&) ()
    at /work1/celeritas/syjun/celeritas/src/celeritas/src/corecel/sys/Device.cc:252
#18 0x00007ffff7ff03b9 in celeritas::SharedParams::Initialize(celeritas::SetupOptions const&) ()
    at /work1/celeritas/syjun/celeritas/src/celeritas/src/accel/SharedParams.cc:62
#19 0x000000000040ed78 in demo_geant::RunAction::BeginOfRunAction(G4Run const*)
    ()
    at /work1/celeritas/syjun/celeritas/src/celeritas/app/demo-geant-integration/RunAction.cc:37

which lead to a memory error at
#4  0x00007ffff7ff20ec in deallocate (
    this=0x7ffff3fa71a8 <celeritas::(anonymous namespace)::global_device()::device+72>, __p=0x7fffbc0a8c80)
    at /srv/software/gnu9/9.3.0/include/c++/9.3.0/bits/stl_tree.h:1911

Reply:
Yeah sounds like @amandalund is right. I'll add a mutex and hopefully that fixes it?